src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Linux.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Sockets
+{
+    sealed class SocketAsyncEngine
+    {
+        private static SocketAsyncEngine _engine;
+        private static readonly object _initLock = new object();
+
+        private readonly int _epollFd;
+
+        public static SocketAsyncEngine Instance
+        {
+            get
+            {
+                if (Volatile.Read(ref _engine) == null)
+                {
+                    lock (_initLock)
+                    {
+                        if (_engine == null)
+                        {
+                            int epollFd = Interop.libc.epoll_create1(Interop.libc.EPOLL_CLOEXEC);
+                            if (epollFd == -1)
+                            {
+                                // TODO: throw an appropriate exception
+                                throw new InternalException();
+                            }
+
+                            var engine = new SocketAsyncEngine(epollFd);
+                            Task.Factory.StartNew(o => {
+                                ((SocketAsyncEngine)o).EventLoop();
+                            }, engine, TaskCreationOptions.LongRunning);
+
+                            Volatile.Write(ref _engine, engine);
+                        }
+                    }
+                }
+
+                return _engine;
+            }
+        }
+
+        private SocketAsyncEngine(int epollFd)
+        {
+            _epollFd = epollFd;
+        }
+
+        private unsafe void EventLoop()
+        {
+            var events = stackalloc Interop.libc.epoll_event[64];
+            for (;;)
+            {
+                int numEvents = Interop.libc.epoll_wait(_epollFd, events, 64, -1);
```

Comment:
As I noted when we spoke yesterday, we should confirm what the behavior is of epoll_wait when there are no file descriptors left in its context.  If it blocks indefinitely, great.  If it immediately wakes up, this has become a busy loop.  something to check.

Additionally, I'm a bit concerned about this thread for the EventLoop sticking around forever, even when there's no work to be done.  It's probably not a big deal: after all, it's only a single thread, and it's rare for an application that uses Sockets to use them only briefly and then never again... more likely is that the application is using Sockets throughout the lifetime of the app.  Something to think about, though.